diff -dPNur boost_1_34_1/boost/date_time/tz_db_base.hpp boost_1_34_1-new/boost/date_time/tz_db_base.hpp
--- boost_1_34_1/boost/date_time/tz_db_base.hpp	2005-10-23 22:15:06.000000000 +0200
+++ boost_1_34_1-new/boost/date_time/tz_db_base.hpp	2008-08-26 04:18:11.000000000 +0200
@@ -158,7 +158,6 @@
       typedef typename time_zone_type::base_type time_zone_base_type;
       typedef typename time_zone_type::time_duration_type time_duration_type;
       typedef time_zone_names_base<char_type> time_zone_names;
-      typedef dst_adjustment_offsets<time_duration_type> dst_adjustment_offsets;
       typedef std::basic_string<char_type> string_type;
 
       //! Constructs an empty database
@@ -346,14 +346,14 @@
         time_duration_type utc_offset = 
           str_from_delimited_time_duration<time_duration_type,char_type>(result[GMTOFFSET]);
         
-        dst_adjustment_offsets adjust(time_duration_type(0,0,0),
+        dst_adjustment_offsets<time_duration_type> adjust(time_duration_type(0,0,0),
                                       time_duration_type(0,0,0),
                                       time_duration_type(0,0,0));
 
         boost::shared_ptr<rule_type> rules;
 
         if(has_dst){
-          adjust = dst_adjustment_offsets(
+          adjust = dst_adjustment_offsets<time_duration_type>(
                                           str_from_delimited_time_duration<time_duration_type,char_type>(result[DSTADJUST]),
                                           str_from_delimited_time_duration<time_duration_type,char_type>(result[START_TIME]),
                                           str_from_delimited_time_duration<time_duration_type,char_type>(result[END_TIME])
